modules/application-server-commons/src/main/java/org/testcontainers/applicationserver/ApplicationServerContainer.java

```diff
@@ -0,0 +1,415 @@
+            withCopyFileToContainer(archive, getApplicationInstallDirectory() + extractApplicationName(archive));
+        }
+
+        // Add lazy env variables
+        for(Map.Entry<String, Supplier<String>> entry : lazyEnvVars.entrySet()) {
+            withEnv(entry.getKey(), entry.getValue().get());
+        }
+    }
+
+    @Override
+    protected void containerIsCreated(String containerId) {
+        if( Objects.isNull(tempDirectory) ) {
+            return;
+        }
+
+        try {
+            //Delete files in temp directory
+            for(String file : tempDirectory.toFile().list()) {
+                Files.deleteIfExists(Paths.get(tempDirectory.toString(), file));
+            }
+            //Delete temp directory
+            Files.deleteIfExists(tempDirectory);
+        } catch (IOException e) {
+            logger().info("Unable to delete temporary directory " + tempDirectory.toString(), e);
+        }
+
+    }
+
+    @Override
+    public void setExposedPorts(List<Integer> exposedPorts) {
+        if( Objects.isNull(this.httpPort) ) {
+            super.setExposedPorts(exposedPorts);
+            return;
+        }
+
+        super.setExposedPorts(appendPort(exposedPorts, this.httpPort));
+    }
+
+    //Configuration
+
+    /**
+     * One or more archives to be deployed to the application platform.
+     *
+     * Calling this method more than once will append new archives to a list to be deployed.
+     *
+     * @param archives - An archive created using shrinkwrap and test runtime
+     * @return self
+     */
+    public ApplicationServerContainer withArchvies(@NonNull Archive<?>... archives) {
+        Stream.of(archives).forEach(archive -> {
+            String name = archive.getName();
+            Path target = Paths.get(getTempDirectory().toString(), name);
+            archive.as(ZipExporter.class).exportTo(target.toFile(), true);
+            this.archives.add(MountableFile.forHostPath(target));
+        });
+
+        return this;
+    }
+
+    /**
+     * One or more archives to be deployed to the application platform.
+     *
+     * Calling this method more than once will append new archives to a list to be deployed.
+     *
+     * @param archives - A MountableFile which represents an archive that was created prior to test runtime.
+     * @return self
+     */
+    public ApplicationServerContainer withArchives(@NonNull MountableFile... archives) {
+        this.archives.addAll(Arrays.asList(archives));
+        return this;
+    }
+
+    /**
+     * This will set the port used to construct the base application URL, as well as
+     * the port used to determine container readiness in the case of Microprofile platforms.
+     *
+     * Calling this method more than once will replace the current httpPort if it was already set.
+     *
+     * @param httpPort - The HTTP port used by the Application platform
+     * @return self
+     */
+    public ApplicationServerContainer withHttpPort(int httpPort) {
```

Comment:
Can you please elaborate more on the use cases for this?

Reply:
This is the port a user would use to connect to their application. 

Similar to the database containers we want to be able to construct a connection URL for testing `getApplicationURL()`.  

It is common for application containers to be configured with a different port than the default.

Example: 
```java
private static ApplicationServerContainer liberty = new LibertyServerContainer(libertyImage)
        .withArchives(createDeployment())
        .withAppContextRoot("test/app/service/")
        .withHttpPort(80);

@Test
public void testURL() {
  String actual = liberty.getApplicationURL();
  String expected = "http://localhost:35286/test/app/service" //where 35286 maps to port 80
  assertThat(actual).isEqualTo(expected);
}
```

You could argue that we could just rely on `firstMappedPort` but for future iterations, I could see us adding in `withHttpsPort` and `getSecureApplicationURL()` which wouldn't be able to relay on that.

Reply:
> It is common for application containers to be configured with a different port than the default.

even for testing? Just curious. I mean, it can be done but I was not expecting it in testing, at least not very often.

Reply:
Yes, if I am an application developer who wants to do integration testing I will have my test application server configured to the same specification as my production application server.  Most applications do not run on the default port `9080`.

Reply:
I suggest to move the method body to the `configure` method. The logic about `replacePort`, `appendPort` seems a little bit odd here. I think we don't need to overwrite `setExposedPorts` neither. 

Tests in `ApplicationServerContainerTest` should call `configure` before `getExposedPorts()` in order to get the proper container configuration. Also, I'm not sure order is required for ports. LMK if I'm missing something

Please, also add tests when readinessPort is set.